File: apps/zap-downloader-node/src/commands/download-zap.ts
import * as path from 'path';
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import chalk from 'chalk';
import { fetchZapVersions } from '../parser';
import { downloadFile, formatBytes } from '../downloader';
import { ZapConfig } from '../types';
import { Arguments } from 'yargs';
import { ensureWorkspace } from '../workspace';

export const command = 'download-zap';
export const describe = 'Download ZAP core and addons from config file';

export const builder = (yargs: any) => {
  return yargs
    .option('config', {
      alias: 'c',
      description: 'Path to ZAP config file',
      type: 'string',
      demandOption: true,
    });
};

export const handler = async (argv: Arguments & {
  config: string;
  workspace: string;
}) => {
  const configPath = argv.config;
  const workspace = argv.workspace;

  if (!fs.existsSync(configPath)) {
    console.error(chalk.red(`Config file not found: ${configPath}`));
    process.exit(1);
  }

  const configContent = fs.readFileSync(configPath, 'utf-8');
  const config: ZapConfig = yaml.load(configContent) as ZapConfig;

  if (!config.zap || !config.addons) {
    console.error(chalk.red('Invalid config: missing zap or addons'));
    process.exit(1);
  }

  ensureWorkspace(workspace);

  console.log(chalk.blue('Fetching ZAP versions...'));
  const zapVersions = await fetchZapVersions();

  const platform = config.zap.platform;
  const version = config.zap.version;

  const platformData = zapVersions.core.platforms[platform as keyof typeof zapVersions.core.platforms];
  if (!platformData) {
    console.error(chalk.red(`Platform ${platform} not available`));
    process.exit(1);
  }

  const zapDir = path.join(workspace, 'zap');
  if (!fs.existsSync(zapDir)) {
    fs.mkdirSync(zapDir, { recursive: true });
  }

  console.log(chalk.yellow(`\nDownloading ZAP core: ${platform} v${version}...`));
  console.log(chalk.gray(`  Size: ${formatBytes(platformData.size)}`));

  const zapOutputPath = path.join(zapDir, platformData.file);
  await downloadFile(platformData.url, zapOutputPath, platformData.hash);
  console.log(chalk.green('ZAP core downloaded!'));

  const addonsDir = path.join(workspace, 'addons');
  if (!fs.existsSync(addonsDir)) {
    fs.mkdirSync(addonsDir, { recursive: true });
  }

  const addonMap = new Map<string, typeof zapVersions.addons[0]>();
  for (const addon of zapVersions.addons) {
    addonMap.set(addon.id, addon);
  }

  const toDownload: typeof zapVersions.addons[0][] = [];
  const downloadedIds = new Set<string>();

  for (const request of config.addons) {
    const addon = addonMap.get(request.id);
    if (!addon) {
      console.warn(chalk.yellow(`Addon not found: ${request.id}`));
      continue;
    }

    if (request.status && addon.status !== request.status) {
      console.warn(chalk.yellow(`Skipping ${addon.id}: status mismatch`));
      continue;
    }

    toDownload.push(addon);
  }

  if (toDownload.length > 0) {
    console.log(chalk.yellow(`\nDownloading ${toDownload.length} addons...`));

    for (const addon of toDownload) {
      if (downloadedIds.has(addon.id)) continue;

      console.log(chalk.gray(`  - ${addon.id} v${addon.version} (${addon.status})`));

      const addonOutputPath = path.join(addonsDir, addon.file);
      await downloadFile(addon.url, addonOutputPath, addon.hash);
      downloadedIds.add(addon.id);
    }
    console.log(chalk.green('Addons downloaded!'));
  }

  console.log(chalk.green('\n=== Download Complete ==='));
  console.log(chalk.blue(`Workspace: ${path.resolve(workspace)}`));
  console.log(chalk.blue(`ZAP: ${path.join(zapDir, platformData.file)}`));
  console.log(chalk.blue(`Addons: ${addonsDir}`));
};

File: apps/zap-downloader/zap_downloader/commands/download_zap.py
import typer
import asyncio
import os
import yaml
from typing import Optional
from rich.console import Console

from ..workspace import ensure_workspace, get_workspace

console = Console()


def download_zap(
    config: str = typer.Option(..., "--config", "-c", help="Path to ZAP config file"),
    workspace: Optional[str] = typer.Option(
        None, "--workspace", "-w", help="Workspace directory"
    ),
    proxy: Optional[str] = typer.Option(None, "--proxy", "-x", help="Proxy URL"),
):
    """Download ZAP core and addons from config file."""
    asyncio.run(_download_zap(config, workspace, proxy))


async def _download_zap(
    config_path: str, workspace: Optional[str], proxy: Optional[str]
):
    from ..parser import fetch_zap_versions
    from ..downloader import download_file, format_bytes

    if not os.path.exists(config_path):
        console.print(f"[red]Config file not found: {config_path}[/red]")
        raise typer.Exit(1)

    with open(config_path, "r") as f:
        config = yaml.safe_load(f)

    if not config.get("zap") or not config.get("addons"):
        console.print("[red]Invalid config: missing zap or addons[/red]")
        raise typer.Exit(1)

    if not workspace:
        workspace = get_workspace()

    ensure_workspace(workspace)

    console.print("[blue]Fetching ZAP versions...[/blue]")
    zap_versions = await fetch_zap_versions(proxy)

    platform = config["zap"]["platform"]
    version = config["zap"]["version"]

    platform_data = zap_versions.core.platforms.get(platform)
    if not platform_data:
        console.print(f"[red]Platform {platform} not available[/red]")
        raise typer.Exit(1)

    zap_dir = os.path.join(workspace, "zap")
    if not os.path.exists(zap_dir):
        os.makedirs(zap_dir, exist_ok=True)

    console.print(f"[yellow]\nDownloading ZAP core: {platform} v{version}...[/yellow]")
    console.print(f"[gray]  Size: {format_bytes(platform_data.size)}[/gray]")

    zap_output_path = os.path.join(zap_dir, platform_data.file)
    await download_file(platform_data.url, zap_output_path, platform_data.hash, proxy)
    console.print("[green]ZAP core downloaded![/green]")

    addons_dir = os.path.join(zap_dir, "addons")
    if not os.path.exists(addons_dir):
        os.makedirs(addons_dir, exist_ok=True)

    addon_map = {addon.id: addon for addon in zap_versions.addons}

    to_download = []
    for request in config["addons"]:
        addon = addon_map.get(request["id"])
        if not addon:
            console.print(f"[yellow]Addon not found: {request['id']}[/yellow]")
            continue

        if request.get("status") and addon.status != request["status"]:
            console.print(f"[yellow]Skipping {addon.id}: status mismatch[/yellow]")
            continue

        to_download.append(addon)

    downloaded_ids = set()
    if to_download:
        console.print(f"[yellow]\nDownloading {len(to_download)} addons...[/yellow]")

        for addon in to_download:
            if addon.id in downloaded_ids:
                continue

            console.print(
                f"[gray]  - {addon.id} v{addon.version} ({addon.status})[/gray]"
            )

            addon_output_path = os.path.join(addons_dir, addon.file)
            await download_file(addon.url, addon_output_path, addon.hash, proxy)
            downloaded_ids.add(addon.id)

        console.print("[green]Addons downloaded![/green]")

    console.print("\n[green]=== Download Complete ===[/green]")
    console.print(f"[blue]Workspace: {os.path.abspath(workspace)}[/blue]")
    console.print(f"[blue]ZAP: {os.path.join(zap_dir, platform_data.file)}[/blue]")
    console.print(f"[blue]Addons: {addons_dir}[/blue]")
